test/RelayHubGasCalculations.test.ts

```diff
@@ -340,6 +348,91 @@ contract('RelayHub gas calculations', function ([_, relayOwner, relayWorker, rel
+      const gassesUsed: any[] = [];
+      [0, 32, 128, 8192/* , 32768, 65536 */].forEach(dataLength => {
+        it(`with arg: ${dynamicArg} length: ${dataLength}`, async function () {
+          // console.log('gasUsed: ', gassesUsed)
+          const id = (await snapshot()).result
+          const senderNonce = (await forwarderInstance.getNonce(senderAddress)).toString()
+          let approvalData = '0x'
+          let paymasterData = '0x'
+          let encodedFunction = recipient.contract.methods.dontEmitMessage('').encodeABI()
+          if (dynamicArg === RelayCallDynamicArg.APPROVAL_DATA) {
+            approvalData = '0x' + 'ff'.repeat(dataLength)
+          } else if (dynamicArg === RelayCallDynamicArg.ENCODED_FUNCTION) {
+            encodedFunction = recipient.contract.methods.dontEmitMessage('f'.repeat(dataLength)).encodeABI()
+            // console.log('encodedFunction', encodedFunction)
+          } else if (dynamicArg === RelayCallDynamicArg.PAYMASTER_DATA) {
+            paymasterData = '0x' + 'ff'.repeat(dataLength)
+          }
+          const relayRequest: RelayRequest = {
+            request: {
+              to: recipient.address,
+              data: encodedFunction,
+              from: senderAddress,
+              nonce: senderNonce,
+              value: '0',
+              gas: gasLimit.toString(),
+              validUntil: '0'
+            },
+            relayData: {
+              baseRelayFee: '0',
+              pctRelayFee: '0',
+              gasPrice: '1',
+              relayWorker,
+              forwarder,
+              paymaster: paymaster.address,
+              paymasterData,
+              clientId
+            }
+          }
+          const dataToSign = new TypedRequestData(
+            chainId,
+            forwarder,
+            relayRequest
+          )
+          const signature = await getEip712Signature(
+            web3,
+            dataToSign
+          )
+          const relayCall = relayHub.contract.methods.relayCall(10e6, relayRequest, signature, approvalData, externalGasLimit)
+          const receipt = await relayCall.send({
+            from: relayWorker,
+            gas: externalGasLimit,
+            gasPrice: gasPrice
+          })
+          gassesUsed.push(receipt.gasUsed - calculateCalldataCost(relayCall.encodeABI()))
+          // console.log('relayCall encodeABI len', relayCall.encodeABI().length / 2)
+          // console.log('gasUsed is', receipt.gasUsed)
+          // console.log('calculateCalldataCost is', calculateCalldataCost(relayCall.encodeABI()))
+          const slack = 2
+          if (gassesUsed.length > 1) {
+            const diff = gassesUsed[gassesUsed.length - 1] - gassesUsed[0]
+            // console.log('diff per byte is', diff / dataLength)
+            // console.log('diff is', diff)
+            const costPerByte = diff / dataLength
```

Comment:
You calculate `diffGas / absoluteDataLength`
you need to keep previous `dataLength` too for a meaningful calculation.

Reply:
@shahafn this sounds correct

Reply:
UPD: `dataLength` is not `absoluteDataLength` as it is just the 'additional' bytes. As this formula is always comparing with 0th gas usage value, it is actually a `diff` in this case